text/1444-union.md

```diff
@@ -178,39 +190,489 @@ fn is_zero(v: Value) -> bool {
+["enum with unknown discriminant"](#towards-better-guarantees-enum-with-unknown-discriminant)
+interpretation of unions.
 
-## Borrowing union fields
+### Contextual keyword
 
-Unsafe code may borrow a reference to a field of a union; doing so borrows the
-entire union, such that any borrow conflicting with a borrow of the union
-(including a borrow of another union field or a borrow of a structure
-containing the union) will produce an error.
+`union` is not a keyword and can be used as identifier in all positions where
+identifiers are permitted. For example:
 
 ```rust
+fn union() {
+    let union = 10;
+    union;
+    union as u8;
+}
+```
+
+`union` is treated as a start of union declaration only if it's found in item
+position (possibly after `pub` and/or attributes) and is followed by a
+non-keyword identifier (union name).  
+This way it doesn't cause any ambiguities in the Rust grammar and can
+be introduced without breaking any existing code that uses `union` as an
+identifier.
+
+### Representation attributes
+
+Attributes `#[repr(C)]` and `#[repr(packed)]` are permitted on unions.
+Attribute `#[repr(simd)]` is permitted on structs, but not on unions.
+
+Layouts of unions using `#[repr(C)]` are usually defined by ABI documents
+provided by C compilers, or hardware vendors, or some standardization
+groups.
+
+Typically, alignment of a union with `#[repr(C)]` is equal to the alignment of
+its most aligned field, and size of a union is equal to the size of its largest
+field rounded up to union alignment (so unions can have trailing padding).  
+Note that those maximums may come from different fields; for instance:
+
+```rust
+#[repr(C)]
 union U {
-    f1: u32,
-    f2: f32,
+    f1: u16,
+    f2: [u8; 4],
+}
+
+fn main() {
+    assert_eq!(std::mem::size_of<U>(), 4);
+    assert_eq!(std::mem::align_of<U>(), 2);
 }
+```
+
+Alignment of a union with `#[repr(C, packed)]` is typically equal to 1, and its
+size is equal to the size of its largest field (so packed unions do not have
+trailing padding).
+
+By default, if `#[repr(C)]` is not used, layout of a union is *unspecified*.
+
+Regardless of representation attributes, all the union fields and the union
+itself are located at the same address in memory (pointers to them are equal).  
+For `#[repr(C)]` unions this property is guaranteed by ISO C standards.
+
+### Dynamically sized fields
+
+Unions cannot contain fields with dynamically sized types like `[u8]` or
+`Trait`.
 
-#[test]
+### Unions and `#[derive]`
+
+Unions support deriving a few traits that do not require accessing their fields.
+Supported traits are
+
+- `Copy`, supported only for unions with `Copy` fields
+- `Clone`, supported only for unions implementing `Copy`, which can be cloned
+trivially
+- `Eq` (but not `PartialEq`), supported for unions with `Eq` fields.
+
+Since accessing union fields reliably requires extra knowledge, traits trying to
+do it (e.g. `PartialEq`) cannot be derived automatically.
```

Comment:
`derive(PartialEq)` need to actually generate code while `derive(Eq)` is just an assertion like `derive(Copy)`. Otherwise https://github.com/rust-lang/rfcs/pull/1897#discussion_r100735217